lib/charms/kratos_external_idp_integrator/v0/kratos_external_provider.py

```diff
@@ -0,0 +1,499 @@
+#!/usr/bin/env python3
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""# Interface library for Kratos external OIDC providers.
+
+This library wraps relation endpoints using the `kratos-external-idp` interface
+and provides a Python API for both requesting Kratos to register the the client credentials for
+communicating with an external provider.
+
+## Getting Started
+
+To get started using the library, you just need to fetch the library using `charmcraft`.
+
+```shell
+cd some-charm
+charmcraft fetch-lib charms.kratos_external_idp_integrator.v0.kratos_external_provider
+```
+
+In the `metadata.yaml` of the charm, add the following:
+
+```yaml
+provides:
+    kratos-external-idp:
+        interface: external_provider
+        limit: 1
+```
+
+Then, to initialise the library:
+
+```python
+from charms.kratos_external_idp_integrator.v0.kratos_external_provider import (
+    ExternalIdpProvider, InvalidConfigError
+)
+from ops.model import BlockedStatus
+
+class SomeCharm(CharmBase):
+  def __init__(self, *args):
+    # ...
+    self.external_idp_provider = ExternalIdpProvider(self, self.config)
+
+    self.framework.observe(self.on.config_changed, self._on_config_changed)
+    self.framework.observe(self.external_idp_provider.on.ready, self._on_ready)
+    self.framework.observe(
+        self.external_idp_provider.on.redirect_uri_changed, self._on_redirect_uri_changed
+    )
+
+    def _on_config_changed(self, event):
+        # ...
+        try:
+            self.external_idp_provider.validate_provider_config(self.config)
+        except InvalidConfigError as e:
+            self.unit.status = BlockedStatus(f"Invalid configuration: {e.args[0]}")
+
+        # ...
+
+    def _on_redirect_uri_changed(self, event):
+        logger.info(f"The client's redirect_uri changed to {event.redirect_uri}")
+        self._stored.redirect_uri = event.redirect_uri
+        self._on_update_status(event)
+
+    def _on_ready(self, event):
+        if not isinstance(self.unit.status, BlockedStatus):
+            self.external_idp_provider.create_provider(self.config)
+```
+"""
+
+import json
+import logging
+
+from ops.framework import EventBase, EventSource, Object, ObjectEvents
+
+# The unique Charmhub library identifier, never change it
+LIBID = "33040051de7f43a8bb43349f2b037dfc"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+DEFAULT_RELATION_NAME = "kratos-external-idp"
+logger = logging.getLogger(__name__)
+
+try:
+    import jsonschema
```

Comment:
Why is it optional? I think it's an important functionality of the library

Reply:
All other relation libraries do it this way, they want to warn the user to install jsonschema without failing. Don't know why.

Reply:
That's weird. I would prefer we enforce validation all the time